src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyFailMessage.java

```diff
@@ -26,18 +31,28 @@
 public class CopyFailMessage extends ControlMessage {
 
   protected static final char IDENTIFIER = 'f';
+  private CopyStatement statement;
 
   private String errorMessage;
 
   public CopyFailMessage(ConnectionHandler connection) throws Exception {
     super(connection);
     this.errorMessage = this.readAll();
+    this.statement = (CopyStatement) connection.getActiveStatement();
   }
 
   @Override
   protected void sendPayload() throws Exception {
-    throw new IllegalStateException(
-        "Spanner does not currently support the copy functionality through the proxy.");
+    // If backend error occurred during copy-in mode, drop any subsequent CopyFail messages.
+    if (!statement.hasException()) {
+      new ErrorResponse(this.outputStream, new Exception(this.errorMessage), State.IOError).send();
+      new ReadyResponse(this.outputStream, ReadyResponse.Status.IDLE).send();
+    } else {
+      MutationWriter mw = this.statement.getMutationWriter();
```

Comment:
Same as above: Shouldn't we send a `ReadyResponse` in this case as well? Put another way: What is the state of the connection if we reach this point, and what is the next message that we expect from the client?

Reply:
Yes, I think you are correct.  We should send a ReadyResponse in this case as well.